src/middlewared/middlewared/plugins/jbof/crud.py

```diff
@@ -0,0 +1,598 @@
+                        # break out of the ifname loop
+                        if node:
+                            self.logger.info(f'Created link on node {node}: {host_ip} -> {shelf_ip}')
+                        else:
+                            self.logger.info(f'Created link: {host_ip} -> {shelf_ip}')
+                        break
+        return list(connected_shelf_ips)
+
+    @private
+    async def attach_drives(self, mgmt_ip, shelf_index, schema, verrors):
+        """Attach drives from the specified expansion shelf."""
+        if await self.middleware.call('failover.licensed'):
+            # HA system
+            if not await self.middleware.call('failover.remote_connected'):
+                verrors.add(schema, 'Unable to contact remote controller')
+                return
+
+            this_node = await self.middleware.call('failover.node')
+            if this_node == 'MANUAL':
+                verrors.add(schema, 'Unable to determine this controllers position in chassis')
+                return
+
+            for node in ('A', 'B'):
+                await self.attach_drives_to_node(node, shelf_index)
+        else:
+            await self.attach_drives_to_node(node, shelf_index)
+
+    @private
+    async def attach_drives_to_node(self, node, shelf_index):
+        localnode = not node or node == await self.middleware.call('failover.node')
+        configured_interfaces = await self.middleware.call('rdma.interface.query')
+        if localnode:
+            for interface in configured_interfaces:
+                if interface['node'] != node:
+                    continue
+                jbof_ip = jbof_static_ip_from_initiator_ip(interface['address'])
+                await self.middleware.call('jbof.nvme_connect', jbof_ip)
+        else:
+            for interface in configured_interfaces:
+                if interface['node'] != node:
+                    continue
+                jbof_ip = jbof_static_ip_from_initiator_ip(interface['address'])
+                try:
+                    await self.middleware.call('failover.call_remote', 'jbof.nvme_connect', [jbof_ip])
+                except CallError as e:
+                    if e.errno != CallError.ENOMETHOD:
+                        raise
+
+    @private
+    def nvme_connect(self, ip, nr_io_queues=16):
+        command = ['nvme', 'connect-all', '-t', 'rdma', '-a', ip, '--persistent', '-i', f'{nr_io_queues}']
+        ret = subprocess.run(command, capture_output=True)
+        if ret.returncode:
+            error = ret.stderr.decode() if ret.stderr else ret.stdout.decode()
+            if not error:
+                error = 'No error message reported'
+            self.logger.debug('Failed to execute command: %r with error: %r', " ".join(command), error)
+            raise CallError(f'Failed connect NVMe disks: {error}')
+        return True
+
+    @private
+    def nvme_disconnect(self, ips):
+        command = ['nvme', 'list-subsys', '-o', 'json']
```

Comment:
Just something that I'm going to note here but not require to be changed in this PR. This command seems to be a bit heavy (and takes a surprising amount of time to return) for the information that we're getting out of it. I was able to write about 10 lines using `pathlib.Path` iterating over sysfs to get this information. I also keyed the returned object on nvme character device name (i.e. `nvme0`, `nvme1` (since those have to be unique)). It would simplify the logic in this function considerably. Anyways, just FYI. If you want the code (I didn't test too thoroughly) just DM me.

Reply:
Updated.